shared/src/transport.rs

```diff
@@ -0,0 +1,97 @@
+use ethcontract::jsonrpc::types::{Call, Value};
+use ethcontract::web3::{error, RequestId, Transport};
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use web3::BatchTransport;
+
+#[derive(Debug, Clone)]
+pub struct LoggingTransport<T: Transport> {
+    inner: T,
+}
+
+impl<T: Transport> LoggingTransport<T> {
+    pub fn new(inner: T) -> LoggingTransport<T> {
+        Self { inner }
+    }
+}
+
+impl<T> Transport for LoggingTransport<T>
+where
+    T: Transport,
+    <T as Transport>::Out: Send + 'static,
+{
+    type Out = BoxFuture<'static, error::Result<Value>>;
+
+    fn prepare(&self, method: &str, params: Vec<Value>) -> (RequestId, Call) {
+        self.inner.prepare(method, params)
+    }
+
+    fn send(&self, id: RequestId, request: Call) -> Self::Out {
+        if let Ok(serialized) = serde_json::to_string(&request) {
+            tracing::debug!("[id:{}] sending request: '{}'", id, &serialized);
+        }
+        self.inner
+            .send(id, request)
+            .map(move |response| {
```

Comment:
ubernit: I think [`inspect`](https://docs.rs/futures/0.3.13/futures/future/trait.FutureExt.html#method.inspect) is slightly more appropriate here as it is for:

> Do something with the output of a future before passing it on.